contracts/GolemNetworkTokenWrapped.sol

```diff
@@ -1,92 +1,39 @@
 pragma solidity ^0.4.16;
 
-import "./ERC223/ERC223.sol";
-import "./ERC223/ERC223ReceivingContract.sol";
-import "./ERC20/ERC20Basic.sol";
-import "./ERC20/ERC20Extended.sol";
+import "./open_zeppelin/StandardToken.sol";
+import "./open_zeppelin/ERC20Basic.sol";
 
-// ERC223 and ERC20 -compliant wrapper token for GNT
-// reworked from original GNTW implementation.
-// Original GNTW implementation notice:
-// // ERC20-compliant wrapper token for GNT
-// // adapted from code provided by u/JonnyLatte
-
-
-// ERC223 is used because it is easier to handle than ERC20.
-// ERC20 is used because of lack of support of ERC223 from Raiden
-// and because ERC223 is in flux
-contract Token is ERC223, ERC20Extended, ERC20Basic {
-
-    function balanceOf(address _owner) view returns (uint256 balance) {
-        return balances[_owner];
-    }
-
-    // Function that is called when a user or another contract wants to transfer funds.
-    function transfer(address _to, uint _value, bytes _data)
-        returns (bool success)
-    {
-        uint codeLength;
-
-        assembly {
-            // Retrieve the size of the code on target address, this needs assembly .
-            codeLength := extcodesize(_to)
-        }
+contract BatchToken is StandardToken {
+    event BatchTransfer(address indexed from, address indexed to, uint256 value,
+        uint64 closureTime);
 
-        var senderBalance = balances[msg.sender];
-        if (senderBalance >= _value && _value > 0) {
-            senderBalance -= _value;
-            balances[msg.sender] = senderBalance;
-            balances[_to] += _value;
-            if(codeLength>0) {
-                ERC223ReceivingContract receiver = ERC223ReceivingContract(_to);
-                // onTokenReceived does revert() if anything goes wrong
-                receiver.onTokenReceived(msg.sender, _value, _data);
-            }
-            // FIXME: when ERC223 will stabilize a bit, revisit this:
-            Transfer(msg.sender, _to, _value);
-            Transfer(msg.sender, _to, _value, _data);
-            return true;
-        }
-        return false;
-    }
+    // This function allows batch payments using sent value and
+    // sender's balance.
+    // Opcode estimation:
+    // Cost: 21000 + 2000 + 5000 + (20000 + 5000) * n
+    // 2000 - arithmetics
+    // 5000 - balance update at the end
+    // 20000 - pesimistic case of balance update in the loop
+    // 5000 - arithmetics + event in the loop
+    function batchTransfer(bytes32[] payments, uint64 closureTime) external {
```

Comment:
Note for later: using `uint64` does not give you anything.

Reply:
Isn't event generation cheaper?
I found a formula (https://www.reddit.com/r/ethdev/comments/78ugki/do_indexed_parameters_in_events_cost_more_gas/) that event generation costs 
```
k + unindexedBytes * a + indexedTopics * b
where
k = 375
a = 8
b = 375
```
and since closureTime is not indexed it would save some gas.

Reply:
That would work if you used `bytes8`. But integer types on ABI level (I assume logs follow the same rules) are extended to the word size (uint256, 32 bytes).